technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+
+* Update status to reflect state of Stargate
+
+#### Event: Reaper ready
+
+##### k8ssandra controller
+
+* Update status to reflect state of Reaper
+
+#### Event: Grafana and Prometheus ready
+
+##### monitoring controller
+
+* Update status of MonitoringConfiguration to reflect state of Grafana and Prometheus
+
+#### Event: MonitoringConfiguration status updated
+
+##### k8ssandra controller
+
+* Update status to reflect state of MonitoringConfiguration
+
+#### Notes
+
+The k8ssandra controller creates several objects:
+
+* CassandraCluster
+* Reaper
+* Stargate
+* MonitoringConfiguration
+
+Whenever any of these objects is updated, it should trigger a reconciliation of the k8ssandra controller. In the reconciliation loop, the status of each object will essentially be pushed up into the K8ssandraCluster object.
+
+### Change Cassandra heap settings
+
+The user updates the heap settings in the K8ssandraCluster manifest and applies the changes with kubectl apply.
+
+#### Event: K8ssandraCluster updated
+
+##### k8ssandra controller
+
+* Update CassandraCluster 
+* End reconciliation loop
+
+#### Event: CassandraCluster updated
+
+##### cassandra cluster controller
+
+* Update CassandraDatacenter(s)
+  * Update CassandraCluster status
+* Requeue reconciliation request
+
+#### Event: CassandraCluster status updated
+
+##### k8ssandra controller
+
+* Update status to reflect state of CassandraCluster
+
+#### Event: CassandraDatacenter update complete and ready
+
+##### cassandra cluster controller
+
+* Update status of CassandraCluster
+* End reconciliation loop
+
+#### Event: CassandraCluster status updated
+
+##### k8ssandra controller
+
+* Update status to reflect state of CassandraCluster
+
+#### Notes
+
+Notice that the user does not directly update the CassandraDatacenter. The CassandraDatacenter is owned and managed by the CassandraCluster controller. Updates made directly to the CassandraDatacenter would potentially be lost. Changes need to be propagated through the K8ssandraCluster object.
+
+## Being Prescriptive
+
+With the Helm charts we are prescriptive with what properties and settings of the CassandraDatacenter we expose. I have been thinking that each parent type would have a TemplateSpec property for each of its child types. For example, K8ssandraCluster would have a CassandraClusterTemplate property that is of type CassandraClusterSpec. CassandraCluster would have something like a []CassandraDatacenterTemplate where CassandraDatacenterTemplate is of type CassandraDatacenterSpec. This provides flexibility and is a pretty common design pattern. 
+
+Aside from deviating away from the current, prescriptive approach there is something else that we have to consider. Suppose the user decides to add a volume or container that clases with one that K8ssandra adds. cass-operator already deals with this with the PodTemplateSpec property. I am not suggesting that we do one thing or another at this time. I just want to raise awareness.
```

Comment:
Can you give an example of what you're talking about here?

Reply:
I wound up swerving into an example in my [previous comment](https://github.com/k8ssandra/k8ssandra/pull/818#discussion_r643960190) but it is worth going over again. Take a look at a StatefulSet or Deployment manifest. Consider this example from the kubernetes [docs](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/):

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
```
The `template` property declare the pod's metadata and spec. Anything that can be declared in a pod manifest (at least the spec), can be declared here. Let's take a look at the underlying Go types to see how this happens.

```go
type DeploymentSpec struct {
     ...

    // Template describes the pods that will be created.
    Template v1.PodTemplateSpec `json:"template" protobuf:"bytes,3,opt,name=template"`

    ...
}

// PodTemplateSpec describes the data a pod should have when created from a template
type PodTemplateSpec struct {
	// Standard object's metadata.
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
	// +optional
	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`

	// Specification of the desired behavior of the pod.
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
	// +optional
	Spec PodSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
}
```

Here is the Go definition for `CassandraDatacenter`:

```go
type CassandraDatacenter struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec   CassandraDatacenterSpec   `json:"spec,omitempty"`
	Status CassandraDatacenterStatus `json:"status,omitempty"`
}
```

The `CassandraCluster`  type definition might look like this:

```go
type CassandraCluster struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec   CassandraClusterSpec   `json:"spec,omitempty"`
	Status CassandraClusterStatus `json:"status,omitempty"`

}

type CassandraClusterSpec struct {
        DatacenterTemplates []CassandraDatacenterTemplateSpec `json:"datacenters"`       
}

type CassandraDatacenterTemplateSpec struct {
        metav1.ObjectMeta `json:"metadata,omitempty"`

        Spec CassandraDatacenterSpec `json:"spec,omitempty"`
}
```

This is not prescriptive like we are today. It provides the flexibility to declare and configure underlying objects as you see fit.